File: 2021/Day 10/C++/part1.cpp


// Headers
#include <algorithm>
#include <array>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <vector>

// Namespaces
using namespace std;  // Bad practice, I know

// Main
int main() {
    const string openBrackets = "([{<";

    // Get input
    ifstream file("../data.txt");
    vector<string> lines;
    string line;
    while (getline(file, line)) lines.push_back(line);

    // Find illegal characters
    map<char, char> bracketMap = {
        {')', '('},
        {']', '['},
        {'}', '{'},
        {'>', '<'},
    };
    vector<char> opens;
    vector<char> illegals;
    for (string line : lines) {
        for (char c : line) {
            if (openBrackets.find(c) != string::npos)
                opens.push_back(c);
            else if (bracketMap[c] == opens.back())
                opens.pop_back();
            else {
                illegals.push_back(c);
                break;
            }
        }
    }

    // Calculate points
    map<char, int> scoringTable = {{')', 3}, {'}', 1197}, {']', 57}, {'>', 25137}};
    int score = 0;
    for (auto c : illegals) score += scoringTable[c];
    cout << "Score: " << score << endl;

    return 0;
}

File: 2021/Day 10/TS/part1.ts
import fs from 'fs'
import path from 'path'

const file: string = fs.readFileSync(path.resolve(__dirname, '../data.txt'), 'utf-8')
const lines = file.split('\n')

// Find illegal characters
const bracketMap: { [key: string]: string } = { ')': '(', '}': '{', ']': '[', '>': '<' }
let chunks: string[] = []
let illegalChars = []
for (let line of lines) {
    for (let c of line) {
        if ('([{<'.includes(c)) chunks.push(c)
        else if (bracketMap[c] == chunks.at(-1)) chunks.pop()
        else {
            illegalChars.push(c)
            break
        }
    }
}

// Calculate points
const scoringTable: { [key: string]: number } = { ')': 3, '}': 1197, ']': 57, '>': 25137 }
const score = illegalChars.reduce((sum, char) => sum + scoringTable[char], 0)
console.log('Score:', score)
